File: Python/modules/dataset.py
from modules.transformations import transformer_superpoint, transformer_ellipseg

import os
import cv2
import torch
import numpy as np



class SuperPointData(torch.utils.data.Dataset):
    def __init__(self, root, df, transform=None, size=None):
        self.root = root
        self.imlist = df.FileName
        self.target = df[['c0x','c0y','cXx','cXy','cYx','cYy']].values
        
        # Transformation for data augmentation
        self.transform = transform
        self.size = size # w,h
    
    def __len__(self):
        return len(self.imlist)
    
    def __getitem__(self, idx):
        im = cv2.imread(os.path.join(self.root,self.imlist[idx]),0)
        target = self.target[idx].copy().reshape(-1,2)
        
        if self.transform:
            im, target = transformer_superpoint(im, target, 0.5, self.size)
        
        # From np.array (HxWxC) to torch.tensor (CxHxW). From [0,255] to [0,1]
        im = torch.from_numpy(np.float32(im/255)).unsqueeze(0)

        # Final mask: {0:origin, 1:x-axis, 2:y-axis, 3:background}
        target = np.int32(np.round(target))
        mask = torch.full(self.size[::-1], 4, dtype=torch.long)
        mask[target[:,1],target[:,0]] = torch.tensor([0,1,2],dtype=torch.long)
        
        return im, mask


class EllipseData(torch.utils.data.Dataset):
    def __init__(self, data, transform, size):
        self.data = data
        self.transform = transform
        self.size = size # w,h
        
    def __len__(self):
        return len(self.data)
    
    def __getitem__(self, idx):
        # Read image and mask
        im = cv2.imread(self.data[idx][0],0)
        mask = cv2.imread(self.data[idx][1],0)
        
        if self.transform:
            im, mask = transformer_ellipseg(im, mask, 0.5, self.size)
        
        # From np.array (HxWxC) to torch.tensor (CxHxW). From [0,255] to [0,1]
        im = torch.from_numpy(np.float32(im/255)).unsqueeze(0)
        mask = torch.from_numpy(np.float32(mask/255)).unsqueeze(0)
        
        return im, mask

File: C++/src/ellipsegnet/dataset.cpp
#include "dataset.hpp"

#include "transformations.hpp" // transformer_ellipseg

#include <opencv2/opencv.hpp>
#include <torch/torch.h>



EllipseData::EllipseData(const Data& _data, bool _transform, cv::Size _sz) : data{_data}, transform{_transform}, sz{_sz} {}

torch::data::Example<> EllipseData::get(size_t idx)
{
    // Read image and mask
    cv::Mat im = cv::imread(data[idx].first,0);
    cv::Mat mask = cv::imread(data[idx].second,0);

    // Data augmentation
    if (transform) transformer_ellipseg(im, mask, 0.5, sz);
    

    // Image from [0,255] to [0,1]
    im.convertTo(im, CV_32F, 1.f/255.f);
    // Convert image to torch tensor
    auto tdata = torch::from_blob(im.data, {im.rows, im.cols, 1}, torch::kF32).clone().permute({2,0,1});

    // Mask from [0,255] to [0,1] and float32
    mask.convertTo(mask, CV_32F, 1.f/255.f);
    // Convert mask to torch tensor
    auto tmask = torch::from_blob(mask.data, {im.rows, im.cols, 1}, torch::kF32).clone().permute({2,0,1});


    return {tdata, tmask};
}

torch::optional<size_t> EllipseData::size() const
{
    return data.size();
}